syntax = "proto3";

package chatroom;

message User {
    string username = 1;
    string password = 2;
}

message UserList {
    string partialusername = 1;
}

message Message {
    string senderusername = 1;
    string receiverusername = 2;
    string message = 3;
}

message requestReply {
    bool status = 1;
    string message = 2;
}

// For server side syncs
message Empty {

}

message internalRequest {
    string command_type = 1;
    string params = 2; // serialized json
    // string user_state = 3;
    // string message_state = 4;
    // string online_state = 5;
}

service ChatRoom {
    rpc CreateUser (User) returns (requestReply);
    rpc Login (User) returns (requestReply);
    rpc Logout (User) returns (requestReply);
    rpc DeleteUser (User) returns (requestReply);
    rpc ListUsers (UserList) returns (requestReply);
    
    rpc SendMessage (Message) returns (requestReply);
    rpc IncomingStream(User) returns (stream Message);
    rpc DeliverMessage(Message) returns (requestReply);
    
    rpc srv_GetNewChanges (Empty) returns (stream internalRequest);
    rpc srv_CheckLeader (Empty) returns (requestReply);
    rpc srv_ElectLeader (Empty) returns (requestReply);
}

// compile using:
// python3 -m grpc_tools.protoc -I../grpc-chatroom --python_out=. --pyi_out=. --grpc_python_out=. ../grpc-chatroom/chatroom.proto